### PR TITLE
Adds simple integration test

### PR DESCRIPTION
**Pull request by xibz**
A very basic integration test that ensures that communication can occur
on the unix socket.

Signed-off-by: xibz <impactbchang@gmail.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -24,4 +24,6 @@ steps:
 
   - label: ':hammer: tests'
     commands:
+      - 'ln -s /var/lib/fc-ci/vmlinux.bin vmlinux'
```
Comment (nmeyerhans) [inline .buildkite/pipeline.yml]: Since you support a `KERNELIMAGE` environment variable, you might as well skip this and just point that variable at `/var/lib/fc-ci/vmlinux.bin`

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -24,4 +24,7 @@ steps:
...
+      - 'touch root-drive.img'
```
Comment (nmeyerhans) [inline .buildkite/pipeline.yml]: Might as well just point the `ROOTDRIVE` environment variable at /dev/null.

File: main_test.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+		rootDrivePath,
+	}
+	cmd := exec.CommandContext(ctlCtx, firectlName, firectlArgs...)
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("Failed to run command: %v", err)
+	}
+
+	defer func() {
+		if err := os.Remove(socketpath); err != nil {
+			t.Log("Failed ot remove socket path file")
```
Comment (nmeyerhans) [inline main_test.go]: typo

File: main_test.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+	// Sleep enough for firecracker to boot up
+	time.Sleep(500 * time.Millisecond)
+	tr := &http.Transport{
+		Dial: func(proto, addr string) (net.Conn, error) {
+			return net.Dial("unix", socketpath)
+		},
+	}
+	client := &http.Client{Transport: tr}
+	_, err := client.Get("http://foo")
```
Comment (nmeyerhans) [inline main_test.go]: It may be overkill, but can you check for a proper InstanceInfo response? You may need to tweak the kernel's `panic` commandline setting to keep the VM from self-destructing long enough for you to do this.

File: main_test.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("Failed to run command: %v", err)
+	}
+
+	defer func() {
+		if err := os.Remove(socketpath); err != nil {
+			t.Log("Failed ot remove socket path file")
+		}
+
+		if err := cmd.Process.Kill(); err != nil {
```
Comment (nmeyerhans) [inline main_test.go]: Might as well send a `SIGQUIT`, after #24 We don't need to request a clean shutdown of this VM.

File: main_test.go (hunk 4 above)
Comment (xibz) [reply main_test.go]: Yea, I was debating whether or not if I wanted to check the response. I'll go ahead and check to ensure that the status is running.

Comment (nmeyerhans) [reply main_test.go]: You can also check that the returned state is something other than "Uninitialized", which indicates that we've successfully started a VM (that is, talked to the API) rather than just started the firecracker process.

File: main_test.go (hunk 5 above)
Comment (xibz) [reply main_test.go]: Why not just rely on the `Kill` method? It sends a `SIGKILL` so I don't think we need to also send a `SIGQUIT`?

File: main_test.go (hunk 4 above)
Comment (xibz) [reply main_test.go]: I am a little more hesitant to check for `Uninitialized` because we have to find the sweet spot of when the VM has been started but not the instance. I think it makes sense to wait for the instance to boot up, then hit it and check for running. What do you think?

File: main_test.go (hunk 5 above)
Comment (nmeyerhans) [reply main_test.go]: We definitely don't want to send a `SIGKILL`. It will terminate the firectl process, but leave an orphaned firecracker process (granted, it should self-destruct once the kernel reboots, but still, we want to let firectl's signal handlers do their thing)

Comment (xibz) [reply main_test.go]: @nmeyerhans - Ah, that makes sense! Cool, will go ahead and signal a `SIGQUIT` instead.

File: main_test.go (hunk 4 above)
Comment (nmeyerhans) [reply main_test.go]: The instance doesn't need to finish booting; we just need to have send the InstanceStart  API, at which point firecracker should return a non-"Uninitialized" value. If you really wanted to, you could retry several times if you keep getting "Uninitialized", and timeout after some reasonably long delay.

File: main_test.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	client := &http.Client{Transport: tr}
+	payload := &models.InstanceInfo{
+		State: models.InstanceInfoStateStarting,
+	}
+
+	// Loop until VM is in a running state
+	for i := 0; i < 10; i++ {
+		if payload.State != models.InstanceInfoStateStarting && payload.State != models.InstanceInfoStateUninitialized {
+			break
```
Comment (nmeyerhans) [inline main_test.go]: Do you actually need to check for `InstanceInfoStateStarting`? If State has a nonzero length and is not `InstanceInfoStateUninitialized`, I think we've seen enough to consider the test a success.

File: main_test.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+		}
+
+		time.Sleep(10 * time.Millisecond)
+		resp, err := client.Get("http://foo")
```
Comment (nmeyerhans) [inline main_test.go]: The logic here feels weird. We don't hit the API endpoint at all until after we've done a comparison for the exit condition (iteration 0 through the for loop).

File: main_test.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		State: models.InstanceInfoStateStarting,
+	}
+
+	// Loop until VM is in a running state
+	for i := 0; i < 10; i++ {
+		if payload.State != models.InstanceInfoStateStarting && payload.State != models.InstanceInfoStateUninitialized {
+			break
+		}
+
+		time.Sleep(10 * time.Millisecond)
```
Comment (nmeyerhans) [inline main_test.go]: In a previous iteration of this PR, you were sleeping for 500 ms before even trying to hit the API endpoint. Now you're failing the test after 100 ms (10 iterations x 10 ms each). Are you sure that's enough (consider that somebody will undoubtedly try to run this on a raspberry pi someday). Maybe you meant to sleep 100 ms?

File: main_test.go [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+			continue
+		}
+
+		payload = &models.InstanceInfo{}
+		if err := json.Unmarshal(b, payload); err != nil {
+			t.Fatalf("Failed to unmarshal response: %v", err)
+		}
+	}
+
+	if payload.State != models.InstanceInfoStateRunning {
```
Comment (nmeyerhans) [inline main_test.go]: I don't know that we need the VM to get all the way to running to consider it a passing test. If we've sent an `InstanceStart` call, that is sufficient IMO.

**Comment by xibz** [comment]
@nmeyerhans - Thanks for the comments! Resolved them.

File: main_test.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+		if err := os.Remove(socketpath); err != nil {
+			t.Log("Failed to remove socket path file")
+		}
+
+		if err := cmd.Process.Signal(syscall.SIGQUIT); err != nil {
+			t.Log("Failed to kill firectl process")
+		}
+	}()
+
+	// Sleep enough for firecracker to boot up
```
Comment (sipsma) [inline main_test.go]: Is this comment still relevant here?

File: main_test.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+		if len(payload.State) != 0 &&
+			payload.State != models.InstanceInfoStateUninitialized {
+			valid = true
+			break
+		}
+	}
+
+	if !valid {
+		t.Errorf("Failed to enter proper state")
```
Comment (sipsma) [inline main_test.go]: If possible, it would help debugging any failures to include what the final (improper) state was in the message here. Alternatively, logging the state in each iteration of the loop above would be equivalent.

File: .buildkite/pipeline.yml [hunk 12]
```diff
@@ -24,4 +24,5 @@ steps:
 
   - label: ':hammer: tests'
     commands:
-      - "make test"
+      - "make"
+      - "KERNELIMAGE=/var/lib/fc-ci/vmlinux.bin ROOTDRIVE=/dev/null make test"
```
Comment (nmeyerhans) [inline .buildkite/pipeline.yml]: Any reason not to make `/dev/null` the default in main_test.go?

File: README.md [hunk 13]
```diff
@@ -106,3 +106,11 @@ Getting Started on AWS
     --root-drive=hello-rootfs.ext4 \
     --kernel-opts="console=ttyS0 noapic reboot=k panic=1 pci=off nomodules rw"
   ```
+
+## Testing
```
Comment (nmeyerhans) [inline README.md]: This file uses "the other style" for marking headings. Please keep it consistent.

File: main_test.go [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+			return net.Dial("unix", socketpath)
+		},
+	}
+	client := &http.Client{Transport: tr}
+
+	valid := false
+	interval := 15 * time.Millisecond
+	for i := 0; i < 5; i++ {
+		time.Sleep(interval)
+		interval *= 2
```
Comment (nmeyerhans) [inline main_test.go]: I'm not sure I'd bother with a backoff technique here. That's useful for avoiding thundering-herd scenarios with service endpoints, but for a local single-tenant service it's not so important. You don't need to change it; I just wanted to point it out.

File: main_test.go (hunk 11 above)
Comment (nmeyerhans) [inline main_test.go]: A more descriptive error message would be helpful. Maybe something like "VM failed to initialize. Can firecracker successfully launch a vm?"

Comment (xibz) [reply main_test.go]: Went with just adding the last state in the error message

File: .buildkite/pipeline.yml (hunk 12 above)
Comment (xibz) [reply .buildkite/pipeline.yml]: That'd make too much sense. Haha, fixed.

File: .buildkite/pipeline.yml [hunk 15]
```diff
@@ -24,4 +24,5 @@ steps:
 
   - label: ':hammer: tests'
     commands:
-      - "make test"
+      - "make"
```
Comment (samuelkarp) [inline .buildkite/pipeline.yml]: ~How come you added this?~

Edit once I read farther: tests require a firectl binary.  Can you add a comment indicating that you're building the binary to use in the tests?

File: README.md [hunk 16]
```diff
@@ -106,3 +106,12 @@ Getting Started on AWS
     --root-drive=hello-rootfs.ext4 \
     --kernel-opts="console=ttyS0 noapic reboot=k panic=1 pci=off nomodules rw"
   ```
+
+Testing
+---
+
+Most tests do not need access to Firecracker or the KVM. However, the
```
Comment (samuelkarp) [inline README.md]: How do I skip the tests that require access to a Firecracker binary or KVM?

File: main_test.go [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+
+var kernelImagePath = "./vmlinux"
+var rootDrivePath = "/dev/null"
+
+func init() {
+	if v := os.Getenv("KERNELIMAGE"); len(v) != 0 {
+		kernelImagePath = v
+	}
+
+	if v := os.Getenv("ROOTDRIVE"); len(v) != 0 {
```
Comment (samuelkarp) [inline main_test.go]: Do you really need the root drive to be overridden if all the tests are fine just using `/dev/null`?

File: main_test.go [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("Failed to run command: %v", err)
+	}
+
+	defer func() {
+		if err := os.Remove(socketpath); err != nil {
+			t.Log("Failed to remove socket path file")
+		}
+
+		if err := cmd.Process.Signal(syscall.SIGQUIT); err != nil {
```
Comment (samuelkarp) [inline main_test.go]: Why `SIGQUIT` and not `SIGTERM`?  `SIGQUIT` usually [requests a core dump](https://www.gnu.org/software/libc/manual/html_node/Termination-Signals.html).

File: main_test.go [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+			return net.Dial("unix", socketpath)
+		},
+	}
+	client := &http.Client{Transport: tr}
+
+	valid := false
+	interval := 50 * time.Millisecond
+	payload := &models.InstanceInfo{}
+	for i := 0; i < 10; i++ {
+		time.Sleep(interval)
```
Comment (samuelkarp) [inline main_test.go]: Do we always want to sleep for 50ms even on the first iteration?

File: main_test.go [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+			continue
+		}
+
+		b, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			t.Log("Failed to read response body:", err)
+			continue
+		}
+
+		payload = &models.InstanceInfo{}
```
Comment (samuelkarp) [inline main_test.go]: Would it be better to use the SDK for this and let it handle the details of creating a client, making the http request, and unmarshaling a response?

Comment (xibz) [reply main_test.go]: Yea, I was on the fence on whether or not to use the SDK, only reason why I didn't is due to the first iteration didn't do anything with the response. I will go ahead and update.

File: main_test.go (hunk 19 above)
Comment (xibz) [reply main_test.go]: Reasoning for this was that the continue block would also need to sleep meaning there would be two locations, one in the if block and another at the end of the for loop. I will go ahead a change it.

File: README.md (hunk 16 above)
Comment (xibz) [reply README.md]: Tests are skip by default if the binary hasn't been compiled and if no kernel image is present. I can update the doc to reflect that or we can change the behavior where you opt in to skipping rather than skipping by default.

File: main_test.go (hunk 18 above)
Comment (xibz) [reply main_test.go]: `SIGQUIT` also cleans up resources, reading up on it a little. My interpretation of `SIGTERM` is if we don't handle that signal, then Firecracker's process may leak.

Comment (samuelkarp) [reply main_test.go]: Not quite.  Both `SIGQUIT` and `SIGTERM` are catch-able signals (see [the man page](http://man7.org/linux/man-pages/man7/signal.7.html)), they just have different conventions and different defaults around them.  If you look at the table describing the signals, you'll see that `SIGQUIT` has a default action of [`Core`](http://man7.org/linux/man-pages/man5/core.5.html), which terminates the process and creates a core dump file, while `SIGTERM` has a default action of `Term` which just terminates the process (cleanly).  The convention is that `SIGTERM` is the signal used to ask a process to gracefully exit (along with `SIGINT`, which is the signal generated when someone presses ctrl-c in a typical terminal) and many programs install handlers specifically for `SIGTERM` so they can gracefully exit.  `SIGQUIT` is less frequently used with a non-default handler, so you may have an unclean exit and a generated core dump file with the contents of memory.

File: README.md (hunk 16 above)
Comment (samuelkarp) [reply README.md]: I'd prefer to have opt-in to skipping.  In general I don't typically like auto-skipped tests as someone may not notice that the test is skipped and then have a false sense of test coverage.

File: main_test.go [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+	if _, err := os.Stat(firectlName); os.IsNotExist(err) {
+		t.Skip("Missing firectl binary")
+	}
+
+	if _, err := os.Stat(kernelImagePath); os.IsNotExist(err) {
+		t.Skip("Missing kernel image")
+	}
+
+	if _, err := os.Stat(rootDrivePath); os.IsNotExist(err) {
+		t.Skip("Missing root drive")
```
Comment (samuelkarp) [inline main_test.go]: This should probably be a failure.  `/dev/null` missing is pretty unexpected.

File: main_test.go [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+func init() {
+	if v := os.Getenv("KERNELIMAGE"); len(v) != 0 {
+		kernelImagePath = v
+	}
+}
+
+func TestFireCTL(t *testing.T) {
+	const firectlName = "./firectl"
+	if _, err := os.Stat(firectlName); os.IsNotExist(err) {
+		t.Skip("Missing firectl binary")
```
Comment (samuelkarp) [inline main_test.go]: As above, I'd prefer that this is `t.Fail` unless the user explicitly skips the test.

File: main_test.go [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	if _, err := os.Stat(kernelImagePath); os.IsNotExist(err) {
+		t.Skip("Missing kernel image")
```
Comment (samuelkarp) [inline main_test.go]: As above, I'd prefer that this is `t.Fail` unless the user explicitly skips the test.

**Comment by xibz** [comment]
@samuelkarp - Fixed up the issues and thanks for the `SIGTERM` vs `SIGQUIT` explanation!

**Comment by nmeyerhans** [comment]
> @samuelkarp - Fixed up the issues and thanks for the `SIGTERM` vs `SIGQUIT` explanation!

Except that is the *default* behavior, but not the behavior of this application. In the signal handlers set up in machine.go (see `installSignalHandlers()`), INT and TERM are translated to CtrlAltDel events to the VM, which may or may not actually translate that event into a VM termination. SIGQUIT actually results in a hard shutdown of firecracker, which is why that's what I suggested we use here.

**Comment by samuelkarp** [comment]
> In the signal handlers set up in machine.go (see installSignalHandlers()), INT and TERM are translated to CtrlAltDel events to the VM, which may or may not actually translate that event into a VM termination. SIGQUIT actually results in a hard shutdown of firecracker, which is why that's what I suggested we use here.

@nmeyerhans That makes sense.  @xibz, SIGQUIT is appropriate to use here, but needs an explanatory comment like the one @nmeyerhans provided.

**Comment by xibz** [comment]
@nmeyerhans and @samuelkarp - Done

File: README.md (hunk 16 above)
Comment (nmeyerhans) [reply README.md]: It looks like you've added a `SKIP_INTEG_TEST` environment setting. Please document it here.

File: main_test.go [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
+
+	"github.com/firecracker-microvm/firecracker-go-sdk"
+	client "github.com/firecracker-microvm/firecracker-go-sdk/client"
+	ops "github.com/firecracker-microvm/firecracker-go-sdk/client/operations"
+	"github.com/go-openapi/strfmt"
+	"github.com/sirupsen/logrus"
+)
+
+const skipIntegrationTest = "SKIP_INTEG_TEST"
```
Comment (samuelkarp) [inline main_test.go]: Can you document this in the README?

File: README.md [hunk 25]
```diff
@@ -106,3 +106,9 @@ Getting Started on AWS
     --root-drive=hello-rootfs.ext4 \
     --kernel-opts="console=ttyS0 noapic reboot=k panic=1 pci=off nomodules rw"
   ```
+
+Testing
+---
+By default the tests require the firectl binary to be built and a kernel image
+to be present. The integration tests look for the binary and kernel image in
+the root directory. To disable these tests simply set `SKIP_INTEG_TEST`.
```
Comment (samuelkarp) [inline README.md]: How come we lost the documentation on `KERNELIMAGE`?

File: README.md [hunk 26]
```diff
@@ -106,3 +106,11 @@ Getting Started on AWS
...
     --kernel-opts="console=ttyS0 noapic reboot=k panic=1 pci=off nomodules rw"
   ```
+
+Testing
+---
+By default the tests require the firectl binary to be built and a kernel image
+to be present. The integration tests look for the binary and kernel image in
+the root directory. By default it will look for vmlinux kernel image. This can
+be overwritten by setting `KERNELIMAGE` to the desired path. To disable these
+tests simply set `SKIP_INTEG_TEST`.
```
Comment (nmeyerhans) [inline README.md]: Because there's a distinction between an **unset** environment variable and a variable that's set to an **empty value**, it's actually important to clarify that `SKIP_INTEG_TEST` needs to be set to an actual value with a nonzero length. Rather than being so wordy, though, I'd just document that the user should set `SKIP_INTEG_TEST=1`.

Also, just to make it clear to the reader, please include a note indicating that these are environment variables.
